gobblin-modules/gobblin-orc/src/main/java/org/apache/gobblin/writer/GobblinBaseOrcWriter.java

```diff
@@ -140,9 +190,17 @@ public State getFinalState() {
   @Override
   public void flush()
       throws IOException {
-    if (rowBatch.size > 0) {
-      orcFileWriter.addRowBatch(rowBatch);
-      rowBatch.reset();
+    try {
+      this.orcFileWriterLock.readLock().lock();
+      if (rowBatch.size > 0) {
+        if (this.orcFileWriter == null) {
```

Comment:
how can this happen? we do lazy init for file writer?

Reply:
The current code does a lazy init iff the estimated record size is not provided in the config. It's not clear to me why we do this.

Reply:
Lazy init is so that we can estimate first before creating the orcFileWriter as it's only able to be created once per file, though I think this check is a relic of me hitting some race condition that should no longer be possible with the re-entrant locks, will take a deeper look

Reply:
Moved all the writer initialization code to the tuneBatchSize() iff the writer is previously null, and since the batchSize defaults to 1000 it should never flush before tuning at least once.